### PR TITLE
Amit climber

### PR DESCRIPTION
**Pull request by AmitBrunicki**

### CODE REVIEW

File: src/main/java/frc/robot/OI.java [hunk 1]
```diff
@@ -1,8 +1,21 @@
 package frc.robot;
 
-public class OI /*GEVALD*/{
+import edu.wpi.first.wpilibj.Joystick;
+
+public class OI /*GEVALD*/ {
+
+    private Joystick left = new Joystick(0);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/OI.java]: delete the content of this class

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 package frc.robot;
 
+import com.spikes2212.command.drivetrains.commands.DriveArcade;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Robot.java]: delete this import

File: src/main/java/frc/robot/RobotMap.java [hunk 3]
```diff
@@ -4,17 +4,30 @@ public class RobotMap {
...
 
+        int DRIVETRAIN_LEFT_SPARKMAX_1 = -1;
+        int DRIVETRAIN_LEFT_SPARKMAX_2 = -1;
+        int DRIVETRAIN_RIGHT_SPARKMAX_1 = -1;
+        int DRIVETRAIN_RIGHT_SPARKMAX_2 = -1;
     }
-    
+
+    public interface PCM {
+        int CLIMBER_FRONT_SOLENOID_1_FORWARD = -1;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java:20]: order this by a set of forward and reverse for each solenoid (forward1, reverse1, forward2, reverse2...)

File: src/main/java/frc/robot/OI.java [hunk 4]
```diff
@@ -1,6 +1,8 @@
 package frc.robot;
 
-public class OI /*GEVALD*/{
+import edu.wpi.first.wpilibj.Joystick;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/OI.java]: remove this import

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package frc.robot.subsystems;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.PneumaticsModuleType;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import frc.robot.RobotMap;
+
+public class Climber extends DashboardedSubsystem {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Climber.java:12]: `CTRL + ALT + L` if needed

File: src/main/java/frc/robot/RobotMap.java (hunk 3 above)
Comment (OfriRosenbaum) [reply src/main/java/frc/robot/RobotMap.java:20]: also, it is called reverse, not backword

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    public InstantCommand backSolenoidOff() {
+        return new InstantCommand(() -> backSolenoid.set(DoubleSolenoid.Value.kReverse));
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putData("front solenoids forward", frontSolenoidsOn());
+        namespace.putData("front solenoids backword", frontSolenoidsOff());
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Climber.java]: reverse, not backword

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        namespace.putData("back solenoids forward", backSolenoidOn());
+        namespace.putData("back solenoids backword", backSolenoidOff());
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Climber.java]: ditto

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Climber.java (hunk 5 above)
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Climber.java:12]: add class documentation

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/OI.java (hunk 1 above)
Comment (AmitBrunicki) [reply src/main/java/frc/robot/OI.java]: OK

File: src/main/java/frc/robot/Robot.java (hunk 2 above)
Comment (AmitBrunicki) [reply src/main/java/frc/robot/Robot.java]: OK

File: src/main/java/frc/robot/RobotMap.java (hunk 3 above)
Comment (AmitBrunicki) [reply src/main/java/frc/robot/RobotMap.java:20]: OK

File: src/main/java/frc/robot/subsystems/Climber.java (hunk 5 above)
Comment (AmitBrunicki) [reply src/main/java/frc/robot/subsystems/Climber.java:12]: Nooooooooooooooooooooooooooooooooooooooooooooooooooooooooo

File: src/main/java/frc/robot/OI.java (hunk 4 above)
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/OI.java]: remove this import

File: src/main/java/frc/robot/RobotMap.java [hunk 8]
```diff
@@ -4,17 +4,30 @@ public class RobotMap {
 
     public interface CAN {
 
+        int DRIVETRAIN_LEFT_SPARKMAX_1 = -1;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java]: delete

File: src/main/java/frc/robot/RobotMap.java [hunk 9]
```diff
@@ -4,17 +4,30 @@ public class RobotMap {
...
+        int DRIVETRAIN_LEFT_SPARKMAX_2 = -1;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java]: ditto

File: src/main/java/frc/robot/RobotMap.java [hunk 10]
```diff
@@ -4,17 +4,30 @@ public class RobotMap {
...
+        int DRIVETRAIN_RIGHT_SPARKMAX_1 = -1;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java]: ditto

File: src/main/java/frc/robot/RobotMap.java [hunk 11]
```diff
@@ -4,17 +4,30 @@ public class RobotMap {
...
+        int DRIVETRAIN_RIGHT_SPARKMAX_2 = -1;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java]: ditto

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
+package frc.robot.subsystems;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.PneumaticsModuleType;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import frc.robot.RobotMap;
+
+/**
+ * Controls the three {@link DoubleSolenoid}s that make the robot climb.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java:10]: A subsystem that..

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+public class Climber extends DashboardedSubsystem {
+
+    private static Climber instance;
+
+    private final DoubleSolenoid frontSolenoid1;
+    private final DoubleSolenoid frontSolenoid2;
+    private final DoubleSolenoid backSolenoid;
+
+    public static Climber getInstance() {
+        if (instance == null)
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: add curly braces

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                            RobotMap.PCM.CLIMBER_FRONT_SOLENOID_1_REVERSE),
+                    new DoubleSolenoid(PneumaticsModuleType.CTREPCM, RobotMap.PCM.CLIMBER_FRONT_SOLENOID_2_FORWARD,
+                            RobotMap.PCM.CLIMBER_FRONT_SOLENOID_2_REVERSE),
+                    new DoubleSolenoid(PneumaticsModuleType.CTREPCM, RobotMap.PCM.CLIMBER_BACK_SOLENOID_FORWARD,
+                            RobotMap.PCM.CLIMBER_BACK_SOLENOID_REVERSE)
+            );
+        return instance;
+    }
+
+    private Climber(String name, DoubleSolenoid frontSolenoid1, DoubleSolenoid frontSolenoid2, DoubleSolenoid backSolenoid) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: namespaceName

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+    public InstantCommand frontSolenoidsOff() {
+        return new InstantCommand(() ->
+        {
+            frontSolenoid1.set(DoubleSolenoid.Value.kReverse);
+            frontSolenoid2.set(DoubleSolenoid.Value.kReverse);
+        });
+    }
+
+    public InstantCommand backSolenoidOn() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: image

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        return new InstantCommand(() -> backSolenoid.set(DoubleSolenoid.Value.kForward));
+    }
+
+    public InstantCommand backSolenoidOff() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: image

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private Climber(String name, DoubleSolenoid frontSolenoid1, DoubleSolenoid frontSolenoid2, DoubleSolenoid backSolenoid) {
+        super(name);
+        this.frontSolenoid1 = frontSolenoid1;
+        this.frontSolenoid2 = frontSolenoid2;
+        this.backSolenoid = backSolenoid;
+    }
+
+    public InstantCommand frontSolenoidsOn() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: change to `openFrontSolenoids()`

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        return new InstantCommand(() ->
+        {
+            frontSolenoid1.set(DoubleSolenoid.Value.kForward);
+            frontSolenoid2.set(DoubleSolenoid.Value.kForward);
+        });
+    }
+
+    public InstantCommand frontSolenoidsOff() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: change to `closeFrontSolenoids()`

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        return new InstantCommand(() -> backSolenoid.set(DoubleSolenoid.Value.kForward));
+    }
+
+    public InstantCommand backSolenoidOff() {
+        return new InstantCommand(() -> backSolenoid.set(DoubleSolenoid.Value.kReverse));
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putData("front solenoids forward", frontSolenoidsOn());
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: change to `open front solenoids`

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+        namespace.putData("front solenoids reverse", frontSolenoidsOff());
+        namespace.putData("back solenoids forward", backSolenoidOn());
+        namespace.putData("back solenoids reverse", backSolenoidOff());
+        namespace.putString("front solenoid 1 value", frontSolenoid1.get()::toString);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java:71]: this should potentially change to lambda expression, not sure yet

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/Climb.java [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.function.Supplier;
+
+public class Climb extends SequentialCommandGroup {
+
+    public static final double DRIVE_SPEED = 0.3;
+
+    private static final RootNamespace namespace = new RootNamespace("climb command");
+
+    private static final Supplier<Double> OPEN_FRONT_SOLENOID_ULTRASONIC_DISTANCE =
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java]: units

File: src/main/java/frc/robot/commands/Climb.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+            namespace.addConstantDouble("open front solenoid ultrasonic distance", 50);
+
+    private static final Supplier<Double> OPEN_BACK_SOLENOID_ULTRASONIC_DISTANCE =
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java]: ditto

File: src/main/java/frc/robot/commands/Climb.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+            namespace.addConstantDouble("open back solenoid ultrasonic distance", 30);
+
+    private static final Supplier<Double> CLOSE_BACK_SOLENOID_ULTRASONIC_DISTANCE =
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java]: ditto

File: src/main/java/frc/robot/commands/Climb.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+            namespace.addConstantDouble("close back solenoid ultrasonic distance", 15);
+
+    private static final Supplier<Double> ULTRASONIC_TOLERANCE =
+            namespace.addConstantDouble("ultrasonic tolerance", 3);
+
+    public Climb() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java]: accept subsystems in constructor, it's more solid

File: src/main/java/frc/robot/commands/Climb.java [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+        Drivetrain drivetrain = Drivetrain.getInstance();
+        Climber climber = Climber.getInstance();
+        addCommands(
+                backAwayFromHab(drivetrain),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java]: First of all, it's hub. Second of all, it should be deck.

File: src/main/java/frc/robot/commands/Climb.java [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private ParallelRaceGroup backAwayFromHab(Drivetrain drivetrain) {
+        return new DriveArcade(drivetrain, -DRIVE_SPEED, 0).withInterrupt(
+                () -> Math.abs(OPEN_FRONT_SOLENOID_ULTRASONIC_DISTANCE.get() - drivetrain.getUltrasonicDistanceInCM())
+                        <= ULTRASONIC_TOLERANCE.get()
+        );
+    }
+
+    private InstantCommand openFrontSolenoid(Climber climber) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java:62]: unnecessary

File: src/main/java/frc/robot/commands/Climb.java [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private ParallelRaceGroup driveForward(Drivetrain drivetrain, Supplier<Double> distanceFromHab) {
+        return new DriveArcade(drivetrain, DRIVE_SPEED, 0).withInterrupt(
+                () -> Math.abs(distanceFromHab.get() - drivetrain.getUltrasonicDistanceInCM()) <=
+                        ULTRASONIC_TOLERANCE.get()
+        );
+    }
+
+    private InstantCommand openBackSolenoid(Climber climber) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java:73]: ditto

File: src/main/java/frc/robot/commands/Climb.java [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+        return climber.openBackSolenoids();
+    }
+
+    private InstantCommand closeFrontSolenoid(Climber climber) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java:77]: ditto

File: src/main/java/frc/robot/commands/Climb.java [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+        return climber.closeFrontSolenoids();
+    }
+
+    private InstantCommand closeBackSolenoid(Climber climber) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java:81]: ditto

File: src/main/java/frc/robot/commands/Climb.java (hunk 21 above)
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/Climb.java]: also document (the other constants too)

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+
+    public InstantCommand closeFrontSolenoids() {
+        return new InstantCommand(() ->
+        {
+            frontSolenoid1.set(DoubleSolenoid.Value.kReverse);
+            frontSolenoid2.set(DoubleSolenoid.Value.kReverse);
+        });
+    }
+
+    public InstantCommand openBackSolenoids() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: solenoid, not solenoids

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+        return new InstantCommand(() -> backSolenoid.set(DoubleSolenoid.Value.kForward));
+    }
+
+    public InstantCommand closeBackSolenoids() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: ditto

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+        return new InstantCommand(() -> backSolenoid.set(DoubleSolenoid.Value.kReverse));
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putData("front solenoids forward", openFrontSolenoids());
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: `open front solenoids`
change for the others too

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+        namespace.putData("front solenoids reverse", closeFrontSolenoids());
+        namespace.putData("back solenoids forward", openBackSolenoids());
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Climber.java]: it's one solenoid

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/Climb.java [hunk 34]
```diff
@@ -0,0 +1,81 @@
...
+
+    public static final double DRIVE_SPEED = 0.3;
+
+    private static final RootNamespace namespace = new RootNamespace("climb command");
+
+    /**
+     * The distance from the deck in which the {@link Climber} front solenoid needs to open.
+     */
+    private static final Supplier<Double> OPEN_FRONT_SOLENOID_ULTRASONIC_IN_CM =
+            namespace.addConstantDouble("open front solenoid ultrasonic in cm", 50);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java]: add a blank line after this line

File: src/main/java/frc/robot/commands/Climb.java [hunk 35]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * The distance from the deck in which the {@link Climber} back solenoid needs to open.
+     */
+    private static final Supplier<Double> OPEN_BACK_SOLENOID_ULTRASONIC_IN_CM =
+            namespace.addConstantDouble("open back solenoid ultrasonic in cm", 30);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java:29]: add a blank line after this line

File: src/main/java/frc/robot/commands/Climb.java [hunk 36]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * The distance from the deck in which the {@link Climber} back solenoid needs to close.
+     */
+    private static final Supplier<Double> CLOSE_BACK_SOLENOID_ULTRASONIC_IN_CM =
+            namespace.addConstantDouble("close back solenoid ultrasonic in cm", 15);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java:35]: add a blank line after this line

File: src/main/java/frc/robot/commands/Climb.java [hunk 37]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * The amount of centimeters that the robot is allowed to miss the target range by.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java]: remove the word range, and change to `the maximum amount...`

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/Climb.java [hunk 38]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.function.Supplier;
+
+public class Climb extends SequentialCommandGroup {
+
+    public static final double DRIVE_SPEED = 0.3;
+
+    private static final RootNamespace namespace = new RootNamespace("climb command");
+
+    /**
+     * The distance from the deck in which the {@link Climber} front solenoid needs to open.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java]: there are two front solenoids, so solenoid should solenoids (and needs should be need)

File: src/main/java/frc/robot/commands/Climb.java [hunk 39]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    private static final Supplier<Double> OPEN_FRONT_SOLENOID_ULTRASONIC_IN_CM =
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java]: there are two front solenoids, so solenoid should solenoids

File: src/main/java/frc/robot/commands/Climb.java [hunk 40]
```diff
@@ -0,0 +1,84 @@
...
+            namespace.addConstantDouble("open front solenoid ultrasonic in cm", 50);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java]: there are two front solenoids, so solenoid should solenoids

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

**Review by OfriRosenbaum — APPROVED** [review]

**Pull request merged**